x-pack/platform/plugins/shared/onechat/public/application/components/agents/edit/tools_selection.tsx

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useMemo } from 'react';
+import {
+  EuiBasicTable,
+  EuiBasicTableColumn,
+  EuiCheckbox,
+  EuiCode,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiLoadingSpinner,
+  EuiSpacer,
+  EuiSwitch,
+  EuiText,
+  EuiTitle,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import type { ToolSelection, ToolDescriptor } from '@kbn/onechat-common';
+import { allToolsSelectionWildcard } from '@kbn/onechat-common';
+import { truncateAtNewline } from '../../../utils/truncate_at_newline';
+
+interface ToolsSelectionProps {
+  tools: ToolDescriptor[];
+  toolsLoading: boolean;
+  selectedTools: ToolSelection[];
+  onToolsChange: (tools: ToolSelection[]) => void;
+  disabled?: boolean;
+}
+
+export const ToolsSelection: React.FC<ToolsSelectionProps> = ({
+  tools,
+  toolsLoading,
+  selectedTools,
+  onToolsChange,
+  disabled = false,
+}) => {
+  // Group tools by provider
+  const toolsByProvider = useMemo(() => {
+    const grouped: Record<string, ToolDescriptor[]> = {};
+    tools.forEach((tool) => {
+      const providerId = tool.meta.providerId;
+      if (!grouped[providerId]) {
+        grouped[providerId] = [];
+      }
+      grouped[providerId].push(tool);
+    });
+    return grouped;
+  }, [tools]);
+
+  const isAllToolsSelectedForProvider = (providerId: string) => {
```

Comment:
suggestion: would be good to pull this out of the react component and have unit tests to validate these scenarios

Reply:
++, let's extract that logic to an util, better for testability (and then add tests, ideally)